### ISSUE TITLE
Autocomplete: Folder name that starts with exclamation mark

### ISSUE THREAD

**Issue by ganego** [chrisant996/clink#511]
Hello,

I'm using clink 1.5.9 on W10.  

Because of how file sorting works, I often use folder names that start with an exclamation mark to pin them to the top.  
I read that `!` is used for history expansion, so I disabled it and restarted clink: `clink set history.expand_mode off`.

But still the TAB autocomplete is not capable to use exclamation marks for folder/file name completion without adding a quotation mark before the them which I find inconvenient.

Can something be done about that or am I overlooking something?

Thank you

**Comment by ganego** [comment]
I don't really know what `!` is for in normal cmd/bat. But maybe only the autocomplete can be enhanced by automatically prepending a `"` before running the autocomplete function? This way it would not interfer with any bat or pasted command and only work with autocomplete when I purposely press TAB.  
Could be hidden behind a config entry.

**Comment by chrisant996** [comment]
I understand the request, I have an idea, and I'm investigating and verifying whether the idea is safe to use.

> I don't really know what `!` is for in normal cmd/bat. 

It's used in the two ways I described in my previous reply.

> But maybe only the autocomplete can be enhanced by automatically prepending a `"` before running the autocomplete function? This way it would not interfer with any bat or pasted command and only work with autocomplete when I purposely press TAB.

The question is not how to quote completions.  That's very clear, and Clink already handles quoting in general.  I described how a solution could be designed that doesn't even need a setting.

The question is what negative side effects will occur if Clink ignores that CMD considers `!` to be a "word break character" in the CMD parsing grammar.  Before making such a change, it's important to assess the impact so it can be taken into account properly.  Messing with word break characters can be dangerous and have undesirable side effects (there have been multiple such regressions from earlier times I've changed the list of word break characters in Clink).

CMD expects that quoting should not be performed automatically on `!`.  This is an arcane technical detail about how parsing works for the CMD language.

What I am going to do is research what else gets broken if Clink were to ignore the fact that `!` is technically a word break character, and how impactful the breakages are.

Depending on that research, maybe it will be reasonably safe to pretend that `!` isn't a word break character, and thus allow it to be treated as part of a completion, and thus get automatically quoted during completion.

But I can't make any guarantees yet because I haven't finished researching it yet.
